File: MiroFlow/web_app/frontend/src/types/task.ts
export type TaskStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';

export interface FileInfo {
  file_id: string;
  file_name: string;
  file_type: string;
  absolute_file_path: string;
}

export interface Task {
  id: string;
  task_description: string;
  config_path: string;
  status: TaskStatus;
  created_at: string;
  updated_at: string;
  current_turn: number;
  max_turns: number;
  step_count: number;
  final_answer: string | null;
  summary: string | null;
  error_message: string | null;
  file_info: FileInfo | null;
  log_path: string | null;
}

export interface TaskCreate {
  task_description: string;
  config_path: string;
  file_id?: string;
}

export interface Message {
  role: string;
  content: string;
}

export interface TaskStatusUpdate {
  id: string;
  status: TaskStatus;
  current_turn: number;
  step_count: number;
  recent_logs: unknown[];
  messages: Message[];
  final_answer: string | null;
  summary: string | null;
  error_message: string | null;
}

export interface TaskListResponse {
  tasks: Task[];
  total: number;
  page: number;
  page_size: number;
}

export interface ConfigListResponse {
  configs: string[];
  default: string;
}

export interface UploadResponse {
  file_id: string;
  file_name: string;
  file_type: string;
  absolute_file_path: string;
}

File: MiroFlow/web_app/models/task.py
# SPDX-FileCopyrightText: 2025 MiromindAI
#
# SPDX-License-Identifier: Apache-2.0

"""Pydantic models for task management."""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


TaskStatus = Literal["pending", "running", "completed", "failed", "cancelled"]


class FileInfo(BaseModel):
    """File information for uploaded files."""

    file_id: str
    file_name: str
    file_type: str
    absolute_file_path: str


class TaskCreate(BaseModel):
    """Request model for creating a new task."""

    task_description: str = Field(
        ..., min_length=1, description="The task/question to process"
    )
    config_path: str = Field(
        default="config/agent_web_demo.yaml", description="Agent config path"
    )
    file_id: str | None = Field(default=None, description="Uploaded file ID")


class TaskResponse(BaseModel):
    """Response model for task data."""

    id: str
    task_description: str
    config_path: str
    status: TaskStatus
    created_at: datetime
    updated_at: datetime

    # Progress info
    current_turn: int = 0
    max_turns: int = 0
    step_count: int = 0

    # Results (populated when completed)
    final_answer: str | None = None
    summary: str | None = None
    error_message: str | None = None

    # File info
    file_info: FileInfo | None = None

    # Log path for debugging
    log_path: str | None = None


class TaskListResponse(BaseModel):
    """Response model for task list."""

    tasks: list[TaskResponse]
    total: int
    page: int
    page_size: int


class Message(BaseModel):
    """Model for LLM conversation message."""

    role: str
    content: str


class TaskStatusUpdate(BaseModel):
    """Model for polling status updates (lightweight)."""

    id: str
    status: TaskStatus
    current_turn: int = 0
    step_count: int = 0
    recent_logs: list[dict[str, Any]] = Field(default_factory=list)
    messages: list[Message] = Field(default_factory=list)
    final_answer: str | None = None
    summary: str | None = None
    error_message: str | None = None


class UploadResponse(BaseModel):
    """Response model for file upload."""

    file_id: str
    file_name: str
    file_type: str
    absolute_file_path: str


class ConfigListResponse(BaseModel):
    """Response model for config list."""

    configs: list[str]
    default: str
